I am using this repository to share some of the analysis methods I used in my previous work and hope this can be useful for others.

## Table of contents
- Nanopore sequencing and analysis
  * [Install Ubuntu, minKNOW, CUDA, and enable gpu based guppy basecalling ](https://github.com/xiaoli-dong/bioinfo_notebook/blob/main/nanopore/linux_and_nanopore.md#linux_and_nanopore)
  * [Install WSL2 on windows 10, and enable CUDA, gpu based guppy basecalling](https://github.com/xiaoli-dong/bioinfo_notebook/blob/main/nanopore/wsl2_and_gpu_guppy.md#wsl2_and_gpu_guppy)
  * [Nanopore genome and metagenome sequence analysis](https://github.com/xiaoli-dong/bioinfo_notebook/blob/main/nanopore/nanopore_data_analysis.md#nanopore_data_analysis)
- Genomics and Metagenomics
  * [ Metagenomics crash course](https://github.com/xiaoli-dong/metagenomics_crash_course)
  * [ Eukaryotic metagenome annotation](https://github.com/xiaoli-dong/bioinfo_notebook/blob/main/eukaryotic/eukaryotic_metagenome_annotation.md#eukaryotic_metagenome_annotation)
- SARS-Cov-2 data analysis
  * [A pipeline used to identify inter and intra variants from illumina covid dataset](https://github.com/xiaoli-dong/bioinfo_notebook/blob/main/sarscov2/covidPipe.sh#covid_pipeline)
